File: backend/src/main/java/de/nak/iaa/sundenbock/dto/commentDTO/CommentDTO.java
package de.nak.iaa.sundenbock.dto.commentDTO;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import de.nak.iaa.sundenbock.dto.userDTO.UserDTO;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;

import java.time.Instant;
import java.util.List;

/**
 * Data Transfer Object representing a comment as exposed by the API.
 * <p>
 * This immutable record is used for reading and transferring comment data across application layers.
 * It includes the comment hierarchy (parent/children), reaction counts, and auditing information
 * such as creation/modification timestamps and users.
 * <p>
 * Jakarta Bean Validation annotations document expected constraints for values handled by the API.
 */
public record CommentDTO(
        Long id,
        @NotNull
        Long ticketId,
        Long parentCommentId,
        @NotBlank(message = "The comment text must not be empty")
        @Size(max = 2000, message = "the comment text must not exceed 2000 characters")
        String commentText,
        int likes,
        int dislikes,
        @JsonIgnoreProperties("childComments")
        List<CommentDTO> childComments,
        //!-- from AuditedEntity --!
        Instant createdDate,
        Instant lastModifiedDate,
        UserDTO createdBy,
        UserDTO lastModifiedBy
        ){}

File: frontend/src/app/features/tickets/data/comment.dto.ts
/**
 * @file comment.dto.ts
 *
 * Definiert die Data Transfer Object (DTO) Interfaces für Kommentare.
 * Dies sind die Rohdatenstrukturen für die API-Kommunikation,
 * z.B. `CommentDto` für gelesene Daten und `CreateCommentDto`
 * für das Erstellen neuer Kommentare.
 */
export interface UserDto {
  id?: number;
  username?: string;
  firstName?: string;
  lastName?: string;
}

export interface CommentDto {
  id?: number;
  ticketId: number;
  parentCommentId?: number;
  commentText: string;
  likes?: number;
  dislikes?: number;
  childComments?: CommentDto[];
  createdDate?: string;       
  lastModifiedDate?: string;  
  createdBy?: UserDto;
  lastModifiedBy?: UserDto;
}

export interface CreateCommentDto {
  ticketId: number;
  parentCommentId?: number;
  commentText: string;
}